src/digraph.rs

```diff
@@ -2265,6 +2278,142 @@ impl PyDiGraph {
+    ///     (but slower).
+    /// :param object obj: The data/weight to associate with the new node.
+    /// :param bool check_cycle: If set to ``True``, validates
+    ///     that the substitution will not introduce cycles before
+    ///     modifying the graph. If set to ``False``, validation is
+    ///     skipped. If not provided, inherits the value
+    ///     of ``check_cycle`` from this instance of
+    ///     :class:`retworkx.PyDiGraph`.
+    /// :param weight_combo_fn: An optional python callable that, when
+    ///     specified, is used to merge parallel edges introduced by the
+    ///     substitution, which will occur when multiple nodes in
+    ///     ``to_replace`` have an incoming edge
+    ///     from the same source node or when multiple nodes in
+    ///     ``to_replace`` have an outgoing edge to the same target node.
+    ///     If this instance of :class:`retworkx.PyDiGraph` is a multigraph,
+    ///     leave this unspecified to preserve parallel edges. If unspecified
+    ///     when not a multigraph, parallel edges and their weights will be
+    ///     combined by choosing one of the edge's weights arbitrarily based
+    ///     on an internal iteration order, subject to change.
+    /// :returns: The index of the newly created node.
+    /// :raises DAGWouldCycle: The cycle check is enabled and the
+    ///     substitution would introduce cycle(s).
+    #[pyo3(
+        text_signature = "(self, to_replace, obj, /, check_cycle=None, weight_combo_fn=None)"
+    )]
+    pub fn substitute_nodes_with_node(
+        &mut self,
+        py: Python,
+        to_replace: Vec<usize>,
+        obj: PyObject,
+        check_cycle: Option<bool>,
+        weight_combo_fn: Option<PyObject>,
+    ) -> PyResult<usize> {
+        let can_contract = |nodes: &IndexSet<NodeIndex>| {
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            let visit_next: Vec<NodeIndex> = nodes
+                .iter()
+                .flat_map(|n| self.graph.edges(*n))
+                .filter_map(|edge| {
+                    let target_node = edge.target();
+                    if !nodes.contains(&target_node) {
+                        Some(target_node)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            // Now, if we can reach any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            let mut dfs = Dfs::from_parts(visit_next, self.graph.visit_map());
+            while let Some(node) = dfs.next(&self.graph) {
+                if nodes.contains(&node) {
+                    // we found a path back to `nodes`
+                    return false;
+                }
+            }
+
+            true
+        };
+
+        let mut indices_to_remove: IndexSet<NodeIndex> =
```

Comment:
I think we can make this a bit faster by using ahash as our hasher (which is what is the default in `hashbrown` and what we use for `DictMap` in the custom return types. It should be a bit faster then the default hasher for `IndexSet`. It is worth confirming this with a quick benchmark though.

```suggestion
        let can_contract = |nodes: &IndexSet<NodeIndex, ahash::RandomState>| {
            // Start with successors of `nodes` that aren't in `nodes` itself.
            let visit_next: Vec<NodeIndex> = nodes
                .iter()
                .flat_map(|n| self.graph.edges(*n))
                .filter_map(|edge| {
                    let target_node = edge.target();
                    if !nodes.contains(&target_node) {
                        Some(target_node)
                    } else {
                        None
                    }
                })
                .collect();

            // Now, if we can reach any of `nodes`, there exists a path from `nodes`
            // back to `nodes` of length > 1, meaning contraction is disallowed.
            let mut dfs = Dfs::from_parts(visit_next, self.graph.visit_map());
            while let Some(node) = dfs.next(&self.graph) {
                if nodes.contains(&node) {
                    // we found a path back to `nodes`
                    return false;
                }
            }

            true
        };

        let mut indices_to_remove: IndexSet<NodeIndex, ahash::RandomState> =
```

Reply:
Is it worth introducing a wrapper type that does this for sets like `DictMap`?

Reply:
It doesn't hurt, but I'm not sure it's really necessary with only one user for right now. If we start using it more it'd be an easier call then.

We did it for `DictMap` because we were using `IndexMap` for all the custom returns to have a deterministic iteration order for everything with a custom return type (which was a lot of places).